source/slang/slang-check-decl.cpp

```diff
@@ -6956,6 +6956,23 @@ namespace Slang
         {
             if (auto calleeDeclRef = as<DeclRefExpr>(resolvedInvoke->functionExpr))
             {
+                // Check that imaginary arguments with lvalue set to true correspond to
+                // parameters that have OutTypeBase attribute.
+                //
+                auto funcType = as<FuncType>(calleeDeclRef->type);
+                for (Index ii = 0; ii < imaginaryArguments.getCount(); ++ii)
+                {
+                    if (imaginaryArguments[ii]->type.isLeftValue)
+                    {
+                        if (!as<OutTypeBase>(funcType->getParamType(ii)))
+                        {
+                            visitor->getSink()->diagnose(
+                                attr, Diagnostics::invalidCustomDerivative);
```

Comment:
We should make the diagnostic more informative.

Reply:
Yeah, all the diagnostics around custom derivatives are a bit opaque at the moment (we only get yes/no feedback)
Let me try to add some information about expected signature.